File: backend/src/main/java/com/se114p12/backend/domain/Product.java
package com.se114p12.backend.domain;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;

@Data
@Entity(name = "products")
@NoArgsConstructor
@AllArgsConstructor
@ToString
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long productId;

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    @JsonBackReference
    private ProductCategory category;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL)
    @JsonManagedReference
    private List<Variation> variations;

    @Column(nullable = false, length = 255)
    @NotBlank
    private String name;

    @Column(length = 500)
    @NotBlank
    private String shortDescription;

    @Column(columnDefinition = "TEXT")
    @NotBlank
    private String detailDescription;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal originalPrice;

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private Instant createdAt;

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private String createdBy;

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private Instant updatedAt;

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private String updatedBy;

    @Column(nullable = false)
    private Boolean isAvailable;
}

File: frontend/MAM/app/src/main/java/com/example/mam/entity/Product.kt
package com.example.mam.entity

import androidx.annotation.DrawableRes
import java.text.DecimalFormat
import java.time.Instant

data class Product(
    val id: String = "",
    val name: String = "",
    val shortDescription: String = "",
    val longDescription: String = "",
    val originalPrice: Int = 0,
    val isAvailable: Boolean = true,
    val idCategory: String = "",
    @DrawableRes val img: Int = 0,
    val createAt: Instant = Instant.now(),
    val updateAt: Instant = Instant.now(),
){
    fun getPriceToString(): String{
        val formatter = DecimalFormat("#,###")
        return "${formatter.format(originalPrice)} VND"
    }
}
